packages/apollo-server/src/utils/ApolloServer.ts

```diff
@@ -0,0 +1,383 @@
+            port: options.port,
+            httpServer: this.http,
+          }),
+          () => s(this.engine.engineListeningAddress),
+        );
+        this.engine.on('error', f);
+        return;
+      }
+
+      this.http.listen(options.port, () => {
+        const la: any = this.http.address();
+        // Convert IPs which mean "any address" (IPv4 or IPv6) into localhost
+        // corresponding loopback ip. Note that the url field we're setting is
+        // primarily for consumption by our test suite. If this heuristic is
+        // wrong for your use case, explicitly specify a frontend host (in the
+        // `frontends.host` field in your engine config, or in the `host`
+        // option to ApolloEngine.listen).
+        let hostForUrl = la.address;
+        if (la.address === '' || la.address === '::') hostForUrl = 'localhost';
+        la.url = `http://${joinHostPort(hostForUrl, la.port)}`;
+        s(la);
+      });
+    });
+  }
+
+  public async stop() {
+    if (this.engine) await this.engine.stop();
+    if (this.http) await new Promise(s => this.http.close(s));
+  }
+
+  private createSubscriptionServer(server: HttpServer, config: ListenOptions) {
+    const { onDisconnect, onConnect, keepAlive } = config;
+    SubscriptionServer.create(
+      {
+        schema: this.schema,
+        execute,
+        subscribe,
+        onConnect: onConnect
+          ? onConnect
+          : (connectionParams: Object) => ({ ...connectionParams }),
+        onDisconnect: onDisconnect,
+        onOperation: async (_: string, connection: ExecutionParams) => {
+          connection.formatResponse = (value: ExecutionResult) => ({
+            ...value,
+            errors: value.errors && value.errors.map(err => formatError(err)),
+          });
+          let context: Context = this.context ? this.context : { connection };
+
+          try {
+            context =
+              typeof this.context === 'function'
+                ? await this.context({ connection })
+                : context;
+          } catch (e) {
+            console.error(e);
+            throw e;
+          }
+
+          return { ...connection, context };
+        },
+        keepAlive,
+      },
+      {
+        server,
+        path: this.graphqlEndpoint,
+      },
+    );
+  }
+
+  async request(request: Request) {
+    if (!this) {
+      throw new Error(`It looks like you tried to call this.request but didn't bind it to the parent class. To fix this,
+when calling this.request, either call it using an error function, or bind it like so:
+
+  this.request.bind(this);
+`);
+    }
+    let context: Context = this.context ? this.context : { request };
+
+    try {
+      context =
+        typeof this.context === 'function'
+          ? await this.context({ req: request })
+          : context;
+    } catch (e) {
+      console.error(e);
+      throw e;
+    }
+
+    return {
+      schema: this.schema,
+      tracing: Boolean(this.engineEnabled),
```

Comment:
OK, for some reason I've been trained to write the latter (and I'd be surprised if V8 doesn't optimize it) — let me know if I should avoid doing that!

Reply:
@glasser These perform equally in recent JS engines and `Boolean` performed marginally better in the past, from what I recall.  I think this now falls into the category of: You do you.

I might argue that `Boolean` has a slight defensive advantage since it mandates use of parenthesis where the `!!` approach could be subject to order of operations in the event that someone were to modify the semantics.  Consider a case where someone appends `* variable` to an existing coercion:

```js
> Boolean(10)
true
> Boolean(10*0)
false
> !!10
true
> !!10*0
0 // Maybe fine, but no longer a Boolean.
> !!(10)
true
> !!(10*0)
false // Thanks, parenthesis.
```

However, a reminder to never use `Boolean` as a constructor:

```js
> typeof new Boolean(true)
'object' // Probably a "whoops".
> typeof Boolean(true)
'boolean'
```